### DISCUSSION TITLE
Why not Lua

### DISCUSSION THREAD

**Discussion by davidwoodburn** [typst/typst#3528] [General]
Why does Typst use its own scripting language rather than use Lua? Why create a new language?

**Comment by frozolotl** [comment]
I think the most important one is that with Lua there wouldn't be a good way of nesting code and markup mode:
```typ
// this is code mode
//  vvvvvvvvvvvvv
   #align(center)[
     I can enter markup _within_ code mode!
   ]
```

Additional features enabled by using our own language:
- Seamless memoization (which is very important for Typst because code is executed multiple times).
- Custom syntax, such as `show` or `set` rules, raw code in code mode, `context` (#3497), labels, and more.
- Automatic joining of expressions:
  ```typ
  #let body = {
    heading[My Heading]
    image("my-image.png")
  }
  // instead of
  #(local body = heading("My Heading") + image("my-image.png"))
  ```

Additionally, if Typst used Lua, it would not be able to support the `#expr` syntax unless it separately parsed Lua itself. Every expression would have to be enclosed in parentheses `#(expr)`, and there would still have to be infrastructure to guarantee that it worked correctly.

Finally, for the convenient markup that Typst promises, we would need to write our own parser and evaluator in any case. Lua could replace code mode, but we would still have to implement markup and math mode.
Integrating Lua wouldn't save much time on the implementation side. I'd say it would even cost us time (even assuming there isn't an existing implementation).

**Reply by davidwoodburn** [reply]
That makes sense. Thank you for the thoughtful reply.
